### PR TITLE
Implement wrapper for HTCondor central manager

### PR DESCRIPTION
**Pull request by tpdownes**
Refactor existing blueprint using vm-instance and startup-script modules using a wrapper that adds the additional functionality of provisioning highly available HTCondor central manager using a managed instance group.

This PR makes the deliberate decision to rely upon MIGs for high availability rather than the "hot-hot" built-in high availability of HTCondor collector daemon and hot-cold HA of negotiator daemon. Reasoning:

1. The hot-hot solutions means you can query either collector for node informations, but their answers may sometimes differ slightly. They are "eventually consistent". This is fine for job scheduling, but a little less great for autoscaling decisions.
2. A MIG provides similar time recovery time for the negotiator daemon (the thing that matches jobs) as the hot-cold failover mechanism.
3. Using a MIG with stateful IP addresses means we do not know the IP addresses of the central managers until after the MIG is provisioned. This creates a cycle within the Terraform graph for the startup-script of the central managers themselves because you need to know the IP address of the other central manager to render it in the configuration file.
We can eliminate the cycle by configuring 1-and-only-1 central manager which is configured with CONDOR_HOST pointing to its own IP address (which is known by the machine and does not need to be rendered in the script).

### Submission Checklist

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cloud HPC Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: community/modules/scheduler/htcondor-central-manager/versions.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    time = {
+      source  = "hashicorp/time"
+      version = "~> 0.9"
+    }
+  }
+  provider_meta "google" {
+    module_name = "blueprints/terraform/hpc-toolkit:htcondor-central-manager/v1.20.0"
+  }
+
+  required_version = ">= 0.13.0"
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-central-manager/versions.tf:32]: nit: I thought we decided on 1.0 as the minimum everywhere. I know we still have stragglers but for new modules I think 1.0 is more appropriate.

File: community/modules/scheduler/htcondor-central-manager/main.tf [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    "ENABLE"  = "TRUE"
+  }
+  enable_oslogin_metadata = var.enable_oslogin == "INHERIT" ? {} : { enable-oslogin = lookup(local.oslogin_api_values, var.enable_oslogin, "") }
+  metadata                = merge(local.network_storage_metadata, local.enable_oslogin_metadata, var.metadata)
+
+  name_prefix = "${var.deployment_name}-cm"
+
+  all_runners = flatten([var.central_manager_runner, local.schedd_runner])
+
+  cm_config = templatefile("${path.module}/templates/condor_config.tftpl", {})
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-central-manager/main.tf:35]: nit: I think the following names are clearer
local.cm_config -> local.cm_config_contents
local.cm_object -> local.cm_object_path

File: community/modules/scheduler/htcondor-central-manager/main.tf [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    postcondition {
+      condition     = self.disk_size_gb <= var.disk_size_gb
+      error_message = "var.disk_size_gb must be set to at least the size of the image (${self.disk_size_gb})"
+    }
+  }
+}
+
+data "google_compute_zones" "available" {
+  project = var.project_id
+  region  = var.region
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-central-manager/main.tf:67]: It seems like there is no option to pick a zone or turn off HA. Does this pose a potential issue for egress?

File: community/modules/scheduler/htcondor-central-manager/main.tf [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  region          = var.region
+  labels          = local.labels
+  deployment_name = var.deployment_name
+
+  runners = local.all_runners
+}
+
+module "central_manager_instance_template" {
+  # tflint-ignore: terraform_module_pinned_source
+  source = "github.com/terraform-google-modules/terraform-google-vm//modules/instance_template?ref=84d7959"
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-central-manager/main.tf:97]: nit: I think I would prefer to use a version ref. Any reason to select a specific commit?

File: community/modules/scheduler/htcondor-central-manager/main.tf [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    response            = ""
+    proxy_header        = "NONE"
+    port                = 9618
+    request             = ""
+    request_path        = ""
+    host                = ""
+    enable_logging      = true
+  }
+
+  update_policy = [{
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-central-manager/main.tf:144]: I think it would be appropriate to document somewhere in layman's terms what the HA characteristics are.

File: community/modules/scheduler/htcondor-central-manager/main.tf [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  oslogin_api_values = {
+    "DISABLE" = "FALSE"
+    "ENABLE"  = "TRUE"
+  }
+  enable_oslogin_metadata = var.enable_oslogin == "INHERIT" ? {} : { enable-oslogin = lookup(local.oslogin_api_values, var.enable_oslogin, "") }
+  metadata                = merge(local.network_storage_metadata, local.enable_oslogin_metadata, var.metadata)
+
+  name_prefix = "${var.deployment_name}-cm"
+
+  all_runners = flatten([var.central_manager_runner, local.schedd_runner])
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-central-manager/main.tf:33]: nit: move this below schedd_runner

File: community/modules/scheduler/htcondor-base/variables.tf [hunk 7]
```diff
@@ -69,9 +69,3 @@ variable "execute_point_roles" {
     "roles/storage.objectViewer",
   ]
 }
-
-variable "central_manager_high_availability" {
```
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-base/variables.tf:73]: Deprecate instead of delete. Message should mention that using module gives HA by default.

File: community/modules/scheduler/htcondor-base/README.md
Comment (nick-stroud) [inline community/modules/scheduler/htcondor-base/README.md:1]: nit: The name `htcondor-base` does not convey much to me and the documentation could use a one line that says what the module is (conceptually) as opposed to just what it does. `htcondor-setup` would make more sense to me I thin.

**Review by nick-stroud — CHANGES_REQUESTED** [review]

File: community/modules/scheduler/htcondor-central-manager/variables.tf [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+variable "subnetwork_self_link" {
+  description = "The self link of the subnetwork in which the HTCondor central manager will be created."
+  type        = string
+  default     = null
+}
+
+variable "instance_image" {
+  description = "Custom VM image with HTCondor and Toolkit support installed."
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/variables.tf:100]: "and Toolkit support installed" means, more or less, the autoscaler virtual environment has been created. I'll change the text to read "Custom VM image with HTCondor installed using the htcondor-install module". We have the main example, but I'm wary about overloading variable descriptions when we already point to the example in the README.

We could do something like add a label to the image with Packer and enforce its existence using a terraform postcondition. There are organizations that enforce what labels are allowed to exist and this might break in those environments.

File: community/modules/scheduler/htcondor-base/README.md
Comment (tpdownes) [reply community/modules/scheduler/htcondor-base/README.md:1]: I'll consider this as a separate PR before merging to develop.

File: community/modules/scheduler/htcondor-central-manager/main.tf (hunk 6 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/main.tf:33]: Done

File: community/modules/scheduler/htcondor-central-manager/README.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+A reference implementation is included in the Toolkit module
+[htcondor-pool-secrets]. You may substitute implementations so long as they
+duplicate the functionality in the references. Usage is demonstrated in the
+[HTCondor example][htc-example].
+
+[htc-example]: ../../../../examples/README.md#htc-htcondoryaml--
+[htcondor-pool-secrets]: ../htcondor-pool-secrets/README.md
+[IDTOKEN]: https://htcondor.readthedocs.io/en/latest/admin-manual/security.html#introducing-idtokens
+
+<!-- BEGINNING OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/README.md:26]: Done

File: community/modules/scheduler/htcondor-central-manager/variables.tf [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    fs_type               = string,
+    mount_options         = string,
+    client_install_runner = map(string)
+    mount_runner          = map(string)
+  }))
+  default = []
+}
+
+variable "disk_size_gb" {
+  description = "Boot disk size in GB"
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/variables.tf:71]: Probably bad. Good catch. I'll change to size of base HPC Rocky 8 image and prevent it from being null. With that, the `postcondition` block is helping us catch bad values.

File: community/modules/scheduler/htcondor-central-manager/variables.tf [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "project_id" {
+  description = "Project in which HTCondor pool will be created"
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/variables.tf:18]: changing to central manager

File: community/modules/scheduler/htcondor-base/variables.tf (hunk 7 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-base/variables.tf:73]: I'd prefer to avoid variable-by-variable deprecations, given that:

- the HTCondor modules are marked experimental
- the htcondor-base module only exists on this branch and is replacing a broader htcondor-configure module on main

I think the most appropriate deprecation-y process here would be to adopt the `checkMovedModules` mechanism with (potentially) expanded guidance beyond "it's the same module with a new name."

File: community/modules/scheduler/htcondor-central-manager/README.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+This module provisions a highly available HTCondor central manager using a [Managed
+Instance Group (MIG)][mig] with auto-healing.
+
+[mig]: https://cloud.google.com/compute/docs/instance-groups
+
+## Usage
+
+Although this provisions an HTCondor central manager with standard configuration,
+for a functioning node, you must supply Toolkit runners as described below:
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/README.md:11]: Addressed

File: community/modules/scheduler/htcondor-central-manager/versions.tf (hunk 1 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/versions.tf:32]: I've updated to 1.1 (which is accurate). My general feeling is that the child modules ought to be accurate, although 1.0 has been out 2 years now so it's not unrealistic to impose that full-stop.

File: community/modules/scheduler/htcondor-central-manager/main.tf (hunk 2 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/main.tf:35]: I'd rather keep the naming alignment across modules consistent than choose one here and another there. I'll add a note to my "debt tracker" as a potential mini-PR.

File: community/modules/scheduler/htcondor-central-manager/main.tf (hunk 3 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/main.tf:68]: I will also create a task to add var.zones to the other 2 infrastructure modules of the HTCondor solution.

File: community/modules/scheduler/htcondor-central-manager/main.tf (hunk 5 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/main.tf:152]: PTAL

File: community/modules/scheduler/htcondor-central-manager/variables.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (nick-stroud) [reply community/modules/scheduler/htcondor-central-manager/variables.tf]: I think this is fine

I am just realizing that you have to build an image. Is this a new restriction. Before (as vm-instance) would we have been able to install via startup-script without building an image.

Either way I am fine with this now that we have deploy command, it just means that the simplest form of the blueprint will always have 3 groups.

File: community/modules/scheduler/htcondor-central-manager/main.tf (hunk 3 above)
Comment (nick-stroud) [reply community/modules/scheduler/htcondor-central-manager/main.tf:68]: LGTM

File: community/modules/scheduler/htcondor-base/variables.tf (hunk 7 above)
Comment (nick-stroud) [reply community/modules/scheduler/htcondor-base/variables.tf:73]: Yes to adding to checkMovedModules.

I will knock var deprecation down to nit but would still advocate for it. Your call.

User flow with old blueprint would be like (var deleted):
- create error: moved module
  - user updates blueprint
- create error: variable does not exist
  - user is confused on next steps??? maybe

User flow with old blueprint would be like (var deprecated):
- create error: moved module
  - user updates blueprint
- deploy error: fyi var has been deprecated, update your blueprint
  - user updates blueprint

**Review by nick-stroud — APPROVED** [review]

File: community/modules/scheduler/htcondor-central-manager/variables.tf (hunk 8 above)
[the remark this answers is not in the document]
Comment (tpdownes) [reply community/modules/scheduler/htcondor-central-manager/variables.tf]: I believe one could provide the install runners to the inputs. It is a bad experience, however, due to the significant setup time and makes autoscaling unimpressive. I think we should steer away from this as much as possible.

I think the 3-group pattern is going to be more common with time.

File: community/modules/scheduler/htcondor-base/variables.tf (hunk 7 above)
Comment (tpdownes) [reply community/modules/scheduler/htcondor-base/variables.tf:73]: Added to checkMovedModules with guidance to read README.md in new location. README.md modified to tell user to read the example and modify blueprint.
